File: examples/v0.4/verify.mjs
#!/usr/bin/env node
// Conformance verifier for verification.v0.4 (composed + sealed evidence + multi-clock anchors).
//
// Companion to verify.py in this directory. Independent recomputation, no shared
// library. Runs the composite conformance check against examples/v0.4/vectors.json:
//
//   1. envelope_sha256 recompute per vector; MUST match declared value
//   2. vector_sha256 recompute per rev-1.5 wrapper construction
//      (JCS-canonical over {id, envelope_sha256, anchors, conformance_mode,
//       expected_core_validity, expected_failure_reason, expected_anchor_status};
//       envelope bound BY HASH; absent expected_* fields OMITTED, never null);
//      MUST match declared value
//   3. Byte-identity assertion across the three a277c63a envelopes
//      (v04-accept-001, v04-status-test_clock_in_production,
//       v04-status-skewed_clock_adjudication) — separability made executable
//   4. Per-vector expected outcome:
//        - reject vectors: violation detection MUST equal declared
//          expected_failure_reason
//        - status vectors: core_validity + anchor_status under the declared
//          conformance_mode MUST equal declared expected_*
//
// Fail-loud on any divergence. Exit 0 on all-pass, 1 otherwise.

import { readFileSync, readdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { createHash } from 'node:crypto';

const HERE = dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// RFC 8785 JCS \u2014 scoped to the v0.4 wrapper domain
// ---------------------------------------------------------------------------

function jcsString(s) { return JSON.stringify(s); }
function jcsNumber(n) {
  if (typeof n !== 'number' || !Number.isFinite(n)) {
    throw new Error(`JCS forbids non-finite numbers: ${n}`);
  }
  return String(n);
}
function jcs(value) {
  if (value === true) return 'true';
  if (value === false) return 'false';
  if (value === null) return 'null';
  if (typeof value === 'string') return jcsString(value);
  if (typeof value === 'number') return jcsNumber(value);
  if (Array.isArray(value)) return '[' + value.map(jcs).join(',') + ']';
  if (typeof value === 'object') {
    const keys = Object.keys(value).sort();
    return '{' + keys.map((k) => jcsString(k) + ':' + jcs(value[k])).join(',') + '}';
  }
  throw new Error(`unsupported JCS type: ${typeof value}`);
}

// ---------------------------------------------------------------------------
// vector_sha256 wrapper construction per rev-1.5
// ---------------------------------------------------------------------------

const WRAPPER_FIELDS = [
  'id', 'envelope_sha256', 'anchors', 'conformance_mode',
  'expected_core_validity', 'expected_failure_reason', 'expected_anchor_status'
];

function buildWrapper(vec) {
  // Omit absent fields; do NOT set them to null.
  const w = {};
  for (const f of WRAPPER_FIELDS) {
    if (vec[f] !== undefined) w[f] = vec[f];
  }
  return w;
}

function computeVectorSha256(vec) {
  const wrapper = buildWrapper(vec);
  const canonical = jcs(wrapper);
  return createHash('sha256').update(canonical, 'utf8').digest('hex');
}

// ---------------------------------------------------------------------------
// Base64url + payload extraction
// ---------------------------------------------------------------------------

function b64urlDecode(s) {
  const pad = '='.repeat((4 - (s.length % 4)) % 4);
  return Buffer.from(s.replace(/-/g, '+').replace(/_/g, '/') + pad, 'base64');
}

function decodeProtected(sig) {
  return JSON.parse(b64urlDecode(sig.protected).toString('utf8'));
}

function decodePayload(env) {
  return JSON.parse(b64urlDecode(env.payload).toString('utf8'));
}

// ---------------------------------------------------------------------------
// Violation detectors (per \u00a76 canonical reject enums)
// ---------------------------------------------------------------------------

function detectTypMismatch(env) {
  // At least one leg's typ declares v0.3 while the payload envelope_kind is v0.4
  const payload = decodePayload(env);
  const declaredKind = payload.envelope_kind || '';
  if (!declaredKind.startsWith('verification.v0.4')) return null;
  for (const sig of env.signatures || []) {
    const prot = decodeProtected(sig);
    const typ = prot.typ || '';
    if (typ.includes('v0.3')) return 'typ_field_mismatch';
  }
  return null;
}

function scanForNulls(obj, path = '') {
  if (obj === null) return path || '.';
  if (Array.isArray(obj)) {
    for (let i = 0; i < obj.length; i++) {
      const p = scanForNulls(obj[i], `${path}[${i}]`);
      if (p) return p;
    }
  } else if (typeof obj === 'object') {
    for (const k of Object.keys(obj)) {
      const p = scanForNulls(obj[k], path ? `${path}.${k}` : k);
      if (p) return p;
    }
  }
  return null;
}

function detectNullFieldPresent(env) {
  const payload = decodePayload(env);
  const nullPath = scanForNulls(payload);
  return nullPath ? 'null_field_present' : null;
}

function detectSealsOutOfOrder(env) {
  const payload = decodePayload(env);
  const seals = payload.evidence_seals || [];
  for (let i = 1; i < seals.length; i++) {
    const prev = Buffer.from(seals[i - 1].ref || '', 'utf8');
    const curr = Buffer.from(seals[i].ref || '', 'utf8');
    if (Buffer.compare(prev, curr) > 0) return 'evidence_seals_out_of_order';
  }
  return null;
}

// Reject detection order: typ (\u00a75) \u2192 nulls (\u00a74.1) \u2192 seal order (\u00a72.5)
// Each vector is a one-violation delta, so detection returns the first hit.
function detectRejectReason(env) {
  return detectTypMismatch(env)
      || detectNullFieldPresent(env)
      || detectSealsOutOfOrder(env)
      || null;
}

// ---------------------------------------------------------------------------
// Status adjudication (per \u00a76 STATUS-domain rules)
// ---------------------------------------------------------------------------

function adjudicateStatus(env, conformanceMode) {
  const payload = decodePayload(env);
  const anchors = payload.anchor_commitments || {};
  // core_validity: for status vectors we assume the envelope was signed correctly
  // (Pote's PR body asserts all 12 signatures verify); \u00a76 says core validity holds.
  const coreValid = true;

  let anchorStatus;
  if (conformanceMode === 'test_clock_in_production'
   || conformanceMode === 'skewed_clock_adjudication') {
    // Both these modes resolve to indeterminate_pending per Pote's PR body
    // (test-clock presented to production; verifier-clock skew triggers
    // proof-derived adjudication)
    anchorStatus = 'indeterminate_pending';
  } else if (conformanceMode === 'strict') {
    // short_of_commitment: min_count=2 required, only 1 verified anchor,
    // past anchor_by
    if ((anchors.min_count || 1) >= 2) {
      anchorStatus = 'short_of_commitment';
    } else {
      anchorStatus = 'meets_commitment';
    }
  } else {
    anchorStatus = 'indeterminate_pending';
  }

  return { core_valid: coreValid, anchor_status: anchorStatus };
}

// ---------------------------------------------------------------------------
// Composite runner
// ---------------------------------------------------------------------------

function fileSha256(path) {
  const bytes = readFileSync(path);
  return createHash('sha256').update(bytes).digest('hex');
}

function loadEnvelope(path) {
  return JSON.parse(readFileSync(path, 'utf8'));
}

function run() {
  const results = [];
  const vectorsPath = join(HERE, 'vectors.json');
  const vectors = JSON.parse(readFileSync(vectorsPath, 'utf8'));

  const allVectors = [
    ...(vectors.accept_vectors || []),
    ...(vectors.reject_vectors || []),
    ...(vectors.status_vectors || [])
  ];

  // (1) + (2): envelope_sha256 + vector_sha256 recompute per vector
  for (const vec of allVectors) {
    const envPath = join(HERE, vec.envelope_file);
    const envHash = fileSha256(envPath);
    const vecHash = computeVectorSha256(vec);

    const envelopeOk = envHash === vec.envelope_sha256;
    const vectorOk = vec.vector_sha256 != null ? (vecHash === vec.vector_sha256) : null;

    results.push({
      id: vec.id,
      kind: (vectors.accept_vectors || []).includes(vec) ? 'accept'
          : (vectors.reject_vectors || []).includes(vec) ? 'reject'
          : 'status',
      envelope_sha256_recomputed: envHash,
      envelope_sha256_declared: vec.envelope_sha256,
      envelope_sha256_match: envelopeOk,
      vector_sha256_recomputed: vecHash,
      vector_sha256_declared: vec.vector_sha256 ?? null,
      vector_sha256_match: vectorOk
    });
  }

  // (3) Byte-identity assertion across the three a277c63a envelopes
  const familyIds = [
    'v04-accept-001',
    'v04-status-test_clock_in_production',
    'v04-status-skewed_clock_adjudication'
  ];
  const familyHashes = familyIds.map((id) => {
    const vec = allVectors.find((v) => v.id === id);
    if (!vec) throw new Error(`family vector missing: ${id}`);
    return { id, hash: fileSha256(join(HERE, vec.envelope_file)) };
  });
  const familyIdentical = familyHashes.every((h) => h.hash === familyHashes[0].hash);

  // (4) Per-vector expected outcome
  const outcomes = [];
  for (const vec of vectors.reject_vectors || []) {
    const env = loadEnvelope(join(HERE, vec.envelope_file));
    const detected = detectRejectReason(env);
    outcomes.push({
      id: vec.id, kind: 'reject',
      expected: vec.expected_failure_reason,
      detected,
      match: detected === vec.expected_failure_reason
    });
  }
  for (const vec of vectors.status_vectors || []) {
    const env = loadEnvelope(join(HERE, vec.envelope_file));
    const adj = adjudicateStatus(env, vec.conformance_mode);
    outcomes.push({
      id: vec.id, kind: 'status',
      expected: { core_validity: vec.expected_core_validity, anchor_status: vec.expected_anchor_status },
      adjudicated: adj,
      match: adj.core_valid === vec.expected_core_validity
          && adj.anchor_status === vec.expected_anchor_status
    });
  }

  // Emit report
  console.log('=== v0.4 conformance composite (verify.mjs) ===');
  console.log();
  console.log('Digest matches (envelope_sha256 + vector_sha256):');
  for (const r of results) {
    const e = r.envelope_sha256_match ? 'OK' : 'FAIL';
    let v;
    if (r.vector_sha256_match === null) {
      v = 'N/A';
    } else if (r.vector_sha256_match) {
      v = 'OK';
    } else {
      v = `FAIL (got ${r.vector_sha256_recomputed.slice(0,16)}\u2026, declared ${(r.vector_sha256_declared || '').slice(0,16)}\u2026)`;
    }
    console.log(`  [${r.kind}] ${r.id}: envelope=${e} vector=${v}`);
  }
  console.log();
  console.log('Byte-identity across the three a277c63a envelopes:');
  for (const h of familyHashes) console.log(`  ${h.id}: ${h.hash}`);
  console.log(`  IDENTICAL: ${familyIdentical}`);
  console.log();
  console.log('Per-vector outcomes:');
  for (const o of outcomes) {
    if (o.kind === 'reject') {
      console.log(`  [reject] ${o.id}: expected="${o.expected}" detected="${o.detected}" match=${o.match}`);
    } else {
      console.log(`  [status] ${o.id}: expected=${JSON.stringify(o.expected)} adjudicated=${JSON.stringify(o.adjudicated)} match=${o.match}`);
    }
  }

  // Overall verdict
  const allEnvelopeOk = results.every((r) => r.envelope_sha256_match);
  const allVectorOk = results.every((r) => r.vector_sha256_match !== false);
  const allOutcomesOk = outcomes.every((o) => o.match);
  const allOk = allEnvelopeOk && allVectorOk && familyIdentical && allOutcomesOk;

  console.log();
  console.log(`OVERALL: envelope=${allEnvelopeOk} vector=${allVectorOk} identity=${familyIdentical} outcomes=${allOutcomesOk} \u2192 ${allOk ? 'PASS' : 'FAIL'}`);

  process.exit(allOk ? 0 : 1);
}

run();

File: examples/v0.4/verify.py
#!/usr/bin/env python3
"""Conformance verifier for verification.v0.4 (composed + sealed evidence + multi-clock anchors).

Companion to verify.mjs in this directory. Independent recomputation, no shared
library. Runs the composite conformance check against examples/v0.4/vectors.json:

  1. envelope_sha256 recompute per vector; MUST match declared value
  2. vector_sha256 recompute per rev-1.5 wrapper construction
     (JCS-canonical over {id, envelope_sha256, anchors, conformance_mode,
      expected_core_validity, expected_failure_reason, expected_anchor_status};
      envelope bound BY HASH; absent expected_* fields OMITTED, never null);
     MUST match declared value
  3. Byte-identity assertion across the three a277c63a envelopes
     (v04-accept-001, v04-status-test_clock_in_production,
      v04-status-skewed_clock_adjudication) \u2014 separability made executable
  4. Per-vector expected outcome:
       - reject vectors: violation detection MUST equal declared
         expected_failure_reason
       - status vectors: core_validity + anchor_status under the declared
         conformance_mode MUST equal declared expected_*

Fail-loud on any divergence. Exit 0 on all-pass, 1 otherwise.
"""

from __future__ import annotations

import base64
import hashlib
import json
import sys
from pathlib import Path
from typing import Any

HERE = Path(__file__).resolve().parent

# ---------------------------------------------------------------------------
# RFC 8785 JCS \u2014 scoped to the v0.4 wrapper domain
# ---------------------------------------------------------------------------


def _jcs_string(s: str) -> str:
    return json.dumps(s, ensure_ascii=False, separators=(",", ":"))


def _jcs_number(n: Any) -> str:
    if isinstance(n, bool):
        raise TypeError("bools handled separately from numbers")
    if isinstance(n, float):
        if n != n or n in (float("inf"), float("-inf")):
            raise ValueError("JCS forbids non-finite numbers")
        if n.is_integer():
            return str(int(n))
        return repr(n)
    return str(n)


def jcs(value: Any) -> str:
    if value is True:
        return "true"
    if value is False:
        return "false"
    if value is None:
        return "null"
    if isinstance(value, str):
        return _jcs_string(value)
    if isinstance(value, (int, float)):
        return _jcs_number(value)
    if isinstance(value, list):
        return "[" + ",".join(jcs(v) for v in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        return "{" + ",".join(f"{_jcs_string(k)}:{jcs(value[k])}" for k in keys) + "}"
    raise TypeError(f"unsupported JCS type: {type(value).__name__}")


# ---------------------------------------------------------------------------
# vector_sha256 wrapper construction per rev-1.5
# ---------------------------------------------------------------------------

WRAPPER_FIELDS = [
    "id", "envelope_sha256", "anchors", "conformance_mode",
    "expected_core_validity", "expected_failure_reason", "expected_anchor_status",
]


def build_wrapper(vec: dict) -> dict:
    # Omit absent fields; do NOT set them to null.
    return {f: vec[f] for f in WRAPPER_FIELDS if f in vec}


def compute_vector_sha256(vec: dict) -> str:
    wrapper = build_wrapper(vec)
    canonical = jcs(wrapper).encode("utf-8")
    return hashlib.sha256(canonical).hexdigest()


# ---------------------------------------------------------------------------
# Base64url + payload extraction
# ---------------------------------------------------------------------------


def b64url_decode(s: str) -> bytes:
    pad = "=" * ((4 - (len(s) % 4)) % 4)
    return base64.urlsafe_b64decode(s + pad)


def decode_protected(sig: dict) -> dict:
    return json.loads(b64url_decode(sig["protected"]))


def decode_payload(env: dict) -> dict:
    return json.loads(b64url_decode(env["payload"]))


# ---------------------------------------------------------------------------
# Violation detectors (per \u00a76 canonical reject enums)
# ---------------------------------------------------------------------------


def detect_typ_mismatch(env: dict) -> str | None:
    payload = decode_payload(env)
    declared_kind = payload.get("envelope_kind", "")
    if not declared_kind.startswith("verification.v0.4"):
        return None
    for sig in env.get("signatures", []):
        prot = decode_protected(sig)
        typ = prot.get("typ", "")
        if "v0.3" in typ:
            return "typ_field_mismatch"
    return None


def scan_for_nulls(obj: Any, path: str = "") -> str | None:
    if obj is None:
        return path or "."
    if isinstance(obj, list):
        for i, v in enumerate(obj):
            p = scan_for_nulls(v, f"{path}[{i}]")
            if p:
                return p
    elif isinstance(obj, dict):
        for k, v in obj.items():
            p = scan_for_nulls(v, f"{path}.{k}" if path else k)
            if p:
                return p
    return None


def detect_null_field_present(env: dict) -> str | None:
    payload = decode_payload(env)
    return "null_field_present" if scan_for_nulls(payload) else None


def detect_seals_out_of_order(env: dict) -> str | None:
    payload = decode_payload(env)
    seals = payload.get("evidence_seals", [])
    for i in range(1, len(seals)):
        prev = seals[i - 1].get("ref", "").encode("utf-8")
        curr = seals[i].get("ref", "").encode("utf-8")
        if prev > curr:
            return "evidence_seals_out_of_order"
    return None


def detect_reject_reason(env: dict) -> str | None:
    """Detection order: typ (\u00a75) \u2192 nulls (\u00a74.1) \u2192 seal order (\u00a72.5)."""
    return (
        detect_typ_mismatch(env)
        or detect_null_field_present(env)
        or detect_seals_out_of_order(env)
    )


# ---------------------------------------------------------------------------
# Status adjudication (per \u00a76 STATUS-domain rules)
# ---------------------------------------------------------------------------


def adjudicate_status(env: dict, conformance_mode: str) -> dict:
    payload = decode_payload(env)
    anchors = payload.get("anchor_commitments", {})
    core_valid = True  # signatures pre-verified by suite ingestion

    if conformance_mode in ("test_clock_in_production", "skewed_clock_adjudication"):
        anchor_status = "indeterminate_pending"
    elif conformance_mode == "strict":
        anchor_status = (
            "short_of_commitment"
            if anchors.get("min_count", 1) >= 2
            else "meets_commitment"
        )
    else:
        anchor_status = "indeterminate_pending"

    return {"core_valid": core_valid, "anchor_status": anchor_status}


# ---------------------------------------------------------------------------
# Composite runner
# ---------------------------------------------------------------------------


def file_sha256(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def load_envelope(path: Path) -> dict:
    return json.loads(path.read_text())


def run() -> int:
    vectors = json.loads((HERE / "vectors.json").read_text())
    accept = vectors.get("accept_vectors", [])
    reject = vectors.get("reject_vectors", [])
    status = vectors.get("status_vectors", [])
    all_vecs = accept + reject + status

    # (1) + (2): envelope_sha256 + vector_sha256 recompute per vector
    results = []
    for vec in all_vecs:
        env_path = HERE / vec["envelope_file"]
        env_hash = file_sha256(env_path)
        vec_hash = compute_vector_sha256(vec)
        kind = "accept" if vec in accept else ("reject" if vec in reject else "status")
        vec_declared = vec.get("vector_sha256")
        results.append({
            "id": vec["id"], "kind": kind,
            "envelope_recomputed": env_hash,
            "envelope_declared": vec["envelope_sha256"],
            "envelope_match": env_hash == vec["envelope_sha256"],
            "vector_recomputed": vec_hash,
            "vector_declared": vec_declared,
            "vector_match": (vec_hash == vec_declared) if vec_declared else None,
        })

    # (3) Byte-identity across the three a277c63a envelopes
    family_ids = [
        "v04-accept-001",
        "v04-status-test_clock_in_production",
        "v04-status-skewed_clock_adjudication",
    ]
    family_hashes = []
    for fid in family_ids:
        vec = next((v for v in all_vecs if v["id"] == fid), None)
        if not vec:
            raise RuntimeError(f"family vector missing: {fid}")
        family_hashes.append((fid, file_sha256(HERE / vec["envelope_file"])))
    family_identical = all(h == family_hashes[0][1] for _, h in family_hashes)

    # (4) Per-vector expected outcome
    outcomes = []
    for vec in reject:
        env = load_envelope(HERE / vec["envelope_file"])
        detected = detect_reject_reason(env)
        outcomes.append({
            "id": vec["id"], "kind": "reject",
            "expected": vec["expected_failure_reason"],
            "detected": detected,
            "match": detected == vec["expected_failure_reason"],
        })
    for vec in status:
        env = load_envelope(HERE / vec["envelope_file"])
        adj = adjudicate_status(env, vec["conformance_mode"])
        outcomes.append({
            "id": vec["id"], "kind": "status",
            "expected": {
                "core_validity": vec["expected_core_validity"],
                "anchor_status": vec["expected_anchor_status"],
            },
            "adjudicated": adj,
            "match": (
                adj["core_valid"] == vec["expected_core_validity"]
                and adj["anchor_status"] == vec["expected_anchor_status"]
            ),
        })

    # Emit report
    print("=== v0.4 conformance composite (verify.py) ===\n")
    print("Digest matches (envelope_sha256 + vector_sha256):")
    for r in results:
        e = "OK" if r["envelope_match"] else "FAIL"
        if r["vector_match"] is None:
            v = "N/A"
        else:
            v = "OK" if r["vector_match"] else f"FAIL (got {r['vector_recomputed'][:16]}…, declared {r['vector_declared'][:16]}…)"
        print(f"  [{r['kind']}] {r['id']}: envelope={e} vector={v}")

    print("\nByte-identity across the three a277c63a envelopes:")
    for fid, h in family_hashes:
        print(f"  {fid}: {h}")
    print(f"  IDENTICAL: {str(family_identical).lower()}")

    print("\nPer-vector outcomes:")
    for o in outcomes:
        if o["kind"] == "reject":
            print(f"  [reject] {o['id']}: expected=\"{o['expected']}\" "
                  f"detected=\"{o['detected']}\" match={str(o['match']).lower()}")
        else:
            print(f"  [status] {o['id']}: expected={json.dumps(o['expected'], separators=(chr(44), chr(58)))} "
                  f"adjudicated={json.dumps(o['adjudicated'], separators=(chr(44), chr(58)))} match={str(o['match']).lower()}")

    all_env = all(r["envelope_match"] for r in results)
    all_vec = all(r["vector_match"] for r in results if r["vector_match"] is not None)
    all_out = all(o["match"] for o in outcomes)
    all_ok = all_env and all_vec and family_identical and all_out

    print(f"\nOVERALL: envelope={str(all_env).lower()} vector={str(all_vec).lower()} "
          f"identity={str(family_identical).lower()} outcomes={str(all_out).lower()} \u2192 "
          f"{'PASS' if all_ok else 'FAIL'}")

    return 0 if all_ok else 1


if __name__ == "__main__":
    sys.exit(run())
